controller/src/main/java/io/pravega/controller/store/stream/records/EpochRecord.java

```diff
@@ -37,25 +38,52 @@
 
     private final int epoch;
     /**
-     * Reference epoch is either the same as epoch or the epoch that originated a chain of duplicates 
-     * that includes this epoch. If we look at it as a graph, then it is a tree of depth one, where 
+     * Reference epoch is either the same as epoch or the epoch that originated a chain of duplicates
+     * that includes this epoch. If we look at it as a graph, then it is a tree of depth one, where
      * the root is the original epoch and the children are duplicates.
      */
     private final int referenceEpoch;
     private final ImmutableList<StreamSegmentRecord> segments;
     private final long creationTime;
     @Getter(AccessLevel.PRIVATE)
     private final Map<Long, StreamSegmentRecord> segmentMap;
-    
+    /**
+     * Number of splits performed as part of scale operation.
+     */
+    private final long splits;
+    /**
+     * Number of merges performed as part of scale operation.
+     */
+    private final long merges;
+
     @Builder
-    public EpochRecord(int epoch, int referenceEpoch, @NonNull ImmutableList<StreamSegmentRecord> segments, long creationTime) {
+    public EpochRecord(int epoch, int referenceEpoch, @NonNull ImmutableList<StreamSegmentRecord> segments,
+                       long creationTime, long splits, long merges) {
         this.epoch = epoch;
         this.referenceEpoch = referenceEpoch;
         this.segments = segments;
         this.creationTime = creationTime;
         this.segmentMap = segments.stream().collect(Collectors.toMap(StreamSegmentRecord::segmentId, x -> x));
+        this.splits = splits;
+        this.merges = merges;
+    }
+
+    /**
+     * Method to calculate number of splits and merges.
+     *
+     * Principle to calculate the number of splits and merges:
+     * 1- An event has occurred if a reference range is present (overlaps) in at least two consecutive target ranges.
+     * 2- If the direction of the check in 1 is forward, then it is a split, otherwise it is a merge.
+     *
+     * @param referenceSegmentsList Reference segment list.
+     * @param targetSegmentsList Target segment list.
+     * @return Number of splits/merges.
+     */
+    public static long findSegmentSplitsMerges(List<StreamSegmentRecord> referenceSegmentsList, List<StreamSegmentRecord> targetSegmentsList) {
```

Comment:
this method doesnt seem fit here as neither of the input params are EpochRecord. we should keep this in its original place.

Reply:
Its a static method, so input params don't need to be members of EpochRecord class.
Its a utility method to find the splits/merges given the segments of earlier and this epoch, I think it fits better here.

Reply:
then it should be in a utility class and not in an unrelated data class (epoch record).
there is a utility class called record helper that can be used for this. 

secondly, moving this method out of persistent stream base should be done if we see this method being used in multiple places.. otherwise it can be a private method there the way it is.

Reply:
fixed